[Swift in WebKit] Create a custom module verifier for libraries in WebKit

rr-codes opened the pull request:
#### 3526fd25329cc96fbe4fabb4d66c940d63f51214
<pre>
[Swift in WebKit] Create a custom module verifier for libraries in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=303716">https://bugs.webkit.org/show_bug.cgi?id=303716</a>
<a href="https://rdar.apple.com/166025196">rdar://166025196</a>

Reviewed by Mike Wyrzykowski.

Since Xcode and SWBuild only support verifying framework modules, create a custom module verifier for
libraries, and use it to verify the bmalloc, WTF, and PAL libraries.

This is enormously beneficial since errors will directly arise for all incorrect files in the verified
library; this contrasts with the current state of affairs where module issues in libraries are only
discovered once one of the &quot;bad&quot; header files are included by a framework which is modularized, and
which will then report the downstream issues, which is basically just playing module whack-a-mole.

The design and logic of this verifier are based on the builtin module verifier of SwiftBuild. The
verifier works by following these steps:

1. The verifier is passed in a JSON list of header files in a TAPI format in a command-line argument,
that correspond to all exposed header files of the library.

2. The verifier then constructs a set of input generation tasks based on the files and the environment to
cover all combinations of verifiable languages and architectures.

3. These tasks are then invoked, each of which creates a Dummy framework that includes all the headers of
the library.

4. For each input generation task, a corresponding module verifier task is then created which forms the
arguments for a clang command.

5. Each module verifier task is invoked, which compiles the implementation file that includes the Dummy
framework, using a set of compiler flags which validate possible module-related issues.

6. The verification passes if the tool exits with a return code of zero, indicating compilation succeeded.

The verifier was designed to be as generic as possible, and also uses most of the same environment
variables as the builtin module verifier for consistency. However, instead of using
`ENABLE_MODULE_VERIFIER`, a new `ENABLE_WK_LIBRARY_MODULE_VERIFIER` variable is used instead, just to
avoid the possibility of accidentally conflicting with the builtin module verifier logic.

For now, `ENABLE_WK_LIBRARY_MODULE_VERIFIER` is set to `NO` for all libraries, since the verifier
revealed several dozen issues across the libraries. Once these are fixed, this flag will be set to `YES`,
preventing any future issues.

Note that this is the recommended solution for verifying library modules, since the builtin module
verifier is designed to only be used with frameworks.

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/Scripts/modules-verifier/GenerateModuleVerifierInputsTask.py: Added.
(_module_verifier_target_set_combinations):
(GenerateModuleVerifierInputsTask):
(GenerateModuleVerifierInputsTask.create_tasks):
(GenerateModuleVerifierInputsTask.__init__):
(GenerateModuleVerifierInputsTask.perform_action):
* Source/WTF/Scripts/modules-verifier/ModuleVerifierLanguage.py: Added.
(LanguageStandard):
(LanguageStandard.c_standards):
(LanguageStandard.cxx_standards):
(Language):
(Language.supported_standards):
(Language.include_statement):
(Language.source_file_extension):
(verify_languages):
* Source/WTF/Scripts/modules-verifier/ModuleVerifierTypes.py: Added.
(ModuleVerifierTarget):
(ModuleVerifierTarget.value):
(ModuleVerifierInputs):
(ModuleVerifierTargetSet):
(ModuleVerifierTargetSet.path_component):
* Source/WTF/Scripts/modules-verifier/VerifyModuleTask.py: Added.
(VerifyModuleTask):
(VerifyModuleTask.__init__):
(VerifyModuleTask.create_command):
(VerifyModuleTask.perform_action):
* Source/WTF/Scripts/modules-verifier/library-modules-verifier.py: Added.
(CommandArguments):
(FileListHeader):
(FileList):
(parse_command_arguments):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/304193@main">https://commits.webkit.org/304193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ba542e39f49232bd368551e6a2642238d922c92

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134806 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46057 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142323 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86713 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/582bcd21-63d3-4adf-b0c0-4971fee8235f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136676 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7095 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103020 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137753 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5546 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120837 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83844 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5382 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2992 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2913 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126844 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114601 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38988 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145019 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133313 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6923 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39564 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6988 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111730 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28354 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5221 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117116 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60856 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6969 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35295 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166216 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6760 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70550 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43442 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6996 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6869 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

adetaylor commented on Source/WTF/Scripts/modules-verifier/library-modules-verifier.py:
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
+
```
This needs a FIXME citing the radar it's working around (i.e. that the standard modules verifier doesn't work with libraries) so that we have a plan to remove all this code once that is fixed.

rr-codes replied:
(addressed over Slack, and also updated commit message a bit for clarity)

rr-codes committed:
[Swift in WebKit] Create a custom module verifier for libraries in WebKit
https://bugs.webkit.org/show_bug.cgi?id=303716
rdar://166025196

Reviewed by Mike Wyrzykowski.

Since Xcode and SWBuild only support verifying framework modules, create a custom module verifier for
libraries, and use it to verify the bmalloc, WTF, and PAL libraries.

This is enormously beneficial since errors will directly arise for all incorrect files in the verified
library; this contrasts with the current state of affairs where module issues in libraries are only
discovered once one of the "bad" header files are included by a framework which is modularized, and
which will then report the downstream issues, which is basically just playing module whack-a-mole.

The design and logic of this verifier are based on the builtin module verifier of SwiftBuild. The
verifier works by following these steps:

1. The verifier is passed in a JSON list of header files in a TAPI format in a command-line argument,
that correspond to all exposed header files of the library.

2. The verifier then constructs a set of input generation tasks based on the files and the environment to
cover all combinations of verifiable languages and architectures.

3. These tasks are then invoked, each of which creates a Dummy framework that includes all the headers of
the library.

4. For each input generation task, a corresponding module verifier task is then created which forms the
arguments for a clang command.

5. Each module verifier task is invoked, which compiles the implementation file that includes the Dummy
framework, using a set of compiler flags which validate possible module-related issues.

6. The verification passes if the tool exits with a return code of zero, indicating compilation succeeded.

The verifier was designed to be as generic as possible, and also uses most of the same environment
variables as the builtin module verifier for consistency. However, instead of using
`ENABLE_MODULE_VERIFIER`, a new `ENABLE_WK_LIBRARY_MODULE_VERIFIER` variable is used instead, just to
avoid the possibility of accidentally conflicting with the builtin module verifier logic.

For now, `ENABLE_WK_LIBRARY_MODULE_VERIFIER` is set to `NO` for all libraries, since the verifier
revealed several dozen issues across the libraries. Once these are fixed, this flag will be set to `YES`,
preventing any future issues.

Note that this is the recommended solution for verifying library modules, since the builtin module
verifier is designed to only be used with frameworks.

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/Scripts/modules-verifier/GenerateModuleVerifierInputsTask.py: Added.
(_module_verifier_target_set_combinations):
(GenerateModuleVerifierInputsTask):
(GenerateModuleVerifierInputsTask.create_tasks):
(GenerateModuleVerifierInputsTask.__init__):
(GenerateModuleVerifierInputsTask.perform_action):
* Source/WTF/Scripts/modules-verifier/ModuleVerifierLanguage.py: Added.
(LanguageStandard):
(LanguageStandard.c_standards):
(LanguageStandard.cxx_standards):
(Language):
(Language.supported_standards):
(Language.include_statement):
(Language.source_file_extension):
(verify_languages):
* Source/WTF/Scripts/modules-verifier/ModuleVerifierTypes.py: Added.
(ModuleVerifierTarget):
(ModuleVerifierTarget.value):
(ModuleVerifierInputs):
(ModuleVerifierTargetSet):
(ModuleVerifierTargetSet.path_component):
* Source/WTF/Scripts/modules-verifier/VerifyModuleTask.py: Added.
(VerifyModuleTask):
(VerifyModuleTask.__init__):
(VerifyModuleTask.create_command):
(VerifyModuleTask.perform_action):
* Source/WTF/Scripts/modules-verifier/library-modules-verifier.py: Added.
(CommandArguments):
(FileListHeader):
(FileList):
(parse_command_arguments):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/304193@main

emw-apple commented on Source/WTF/Scripts/modules-verifier/library-modules-verifier.py:
```diff
@@ -0,0 +1,104 @@
+
+
+if __name__ == "__main__":
+    if os.environ.get("ENABLE_WK_LIBRARY_MODULE_VERIFIER") != "YES":
+        print(
+            "warning: Library module verifier is not enabled. Set `ENABLE_WK_LIBRARY_MODULE_VERIFIER` to `YES` to enable the verifier."
+        )
+        sys.exit()
```
This shouldn't be a warning, just a silent exit, unless you mean to say that targets which have the script phase but don't set ENABLE_WK_LIBRARY_MODULE_VERIFIER are _always_ misconfigured. It seems like we might always have configurations that would wish to turn off verification (such as waiting on SDK changes, build performance, new platform bringup).

emw-apple commented on Source/WTF/Scripts/modules-verifier/library-modules-verifier.py:
```diff
@@ -0,0 +1,104 @@
+        sys.exit()
+
+    arguments = parse_command_arguments()
+
+    assert arguments.tapi_filelist.is_file()
+    assert arguments.relative_to.is_dir()
+
```
python nit: These seem unnecessary; the tool will fail immediately if they arguments aren't passed the way they are described.

emw-apple commented on Source/WTF/Scripts/modules-verifier/library-modules-verifier.py:
```diff
@@ -0,0 +1,104 @@
+    for input_task in input_tasks:
+        input_task.perform_action(headers)
+
+    print("Generated inputs for module verifier!")
```
This seems like a debugging line that shouldn't be printed in every successful run.

emw-apple commented on Source/WTF/Scripts/modules-verifier/library-modules-verifier.py:
```diff
@@ -0,0 +1,104 @@
+        command = verify_task.create_command()
+
+        print(
+            f"Verifying clang module ({verify_task.language.value}, {verify_task.standard.value}, {verify_task.target}) ..."
```
Ditto, maybe? Does Xcode's module verifier print something like this too?

emw-apple commented on Source/WTF/Scripts/modules-verifier/library-modules-verifier.py:
```diff
@@ -0,0 +1,104 @@
+        if result.returncode:
+            sys.exit("error: Failed to verify module.")
+
+    print("Verified module!")
```
Ditto.

emw-apple commented on Source/WTF/Scripts/modules-verifier/ModuleVerifierTypes.py:
```diff
@@ -0,0 +1,34 @@
+    os_version: str
+    suffix: Optional[str]
+
+    def value(self) -> str:
```
python nit: you might prefer to call this `__str__` and make it the standard string representation.

emw-apple reviewed (commented):
Leaving some post-hoc review. Overall seems fine, though I am disappointed we are not considering it a bug that Xcode's module verifier does not support library modules.
